### ISSUE TITLE
Docker image doesn't run on (older) Synology NAS)

### ISSUE THREAD

**Issue by JOJ0** [ampache/ampache-docker#114]
Synology NAS systems aged 10+ years old run a kernel 3.x.something, the Ampache docker image uses a compiled apache2 package that requires features (urandom related) that these kernels don't have.

I suggest providing a docker image that is compatible with such older machines/kernels. Updating kernels ony such NAS boxes is not a good idea I suppose.

I think the best stolution would be to create an ampacher container that is based on Debian buster, instead of bullsyeye. I'm trying to get one built with an altered `Dockerfile` at the moment but no success so far.

Drop me a note for any hints with similar problems.

HTH

**Comment by JOJ0** [comment]
This is the specific error when hooked up to a container shell and trying to run Apache2:

```
root@1c23aa7acb77:/# service apache2 start
Starting Apache httpd web server: apache2 failed!
The apache2 configtest failed. ... (warning).
Output of config test was:
[Sun Mar 10 12:16:55.917400 2024] [:crit] [pid 530] (38)Function not implemented: AH00141: Could not initialize random number generator
Action 'configtest' failed.
The Apache error log may have more information.
root@1c23aa7acb77:/#

**Comment by JOJ0** [comment]
Update: I can run older Docker images: The newest that does not have this issue (probably still be Debian Buster), is

```
 image: ampache/ampache:4.4.3
```

**Comment by mitchray** [comment]
I'm a docker (and linux) noob so forgive me, I thought everything is self-contained so why does the kernel matter?

**Comment by JOJ0** [comment]
> I'm a docker (and linux) noob so forgive me, I thought everything is self-contained so why does the kernel matter?

Nope, exactly not the case. To put it simple, the difference between of "full virtualization" (as in a fully virtualized computer, what we call simply a "Virtual Machine/VM") has its _own_ kernel, whereas a container shares the kernel with the host machine.

Hope that clarifies and educates :-)

**Comment by JOJ0** [comment]
But this brings me to a new idea: The problem seems to be Apache included in Debian bullsyeye that requires that kernel feature......

Ok, so is there maybe a Docker Ampache image out there that does not use Apache but Nginx as the webserver (or even any other webserver)?

**Comment by kuzi-moto** [comment]
According to this [SO post](https://stackoverflow.com/questions/66617147/how-to-fix-38function-not-implemented-ah00141-could-not-initialize-random-n), the last working version of Apache that didn't require this kernel feature is 2.4.18.

You could modify the Ampache Dockerfile to manually install that version of Apache. This isn't a great solution as it requires you to run vulnerable software.

As you mentioned using a different webserver altogether is a much better idea. There is no "official" image that uses Nginx, but I did find this repository from 6 years ago that you could try: https://github.com/OnekO/ampache-alpine-nginx/tree/master You'll want to set a couple of the environment variables found in the Dockerfile likely.

If that doesn't work, it should be possible to modify the Ampache Dockerfile to replace Apache with Nginx. You would just need to make sure to include the relevant Nginx configuration. I might be able to take a stab at it this week if I can find some time but no promises.

**Comment by lachlan-00** [comment]
> According to this [SO post](https://stackoverflow.com/questions/66617147/how-to-fix-38function-not-implemented-ah00141-could-not-initialize-random-n), the last working version of Apache that didn't require this kernel feature is 2.4.18.
> 
> You could modify the Ampache Dockerfile to manually install that version of Apache. This isn't a great solution as it requires you to run vulnerable software.
> 
> As you mentioned using a different webserver altogether is a much better idea. There is no "official" image that uses Nginx, but I did find this repository from 6 years ago that you could try: https://github.com/OnekO/ampache-alpine-nginx/tree/master You'll want to set a couple of the environment variables found in the Dockerfile likely.
> 
> If that doesn't work, it should be possible to modify the Ampache Dockerfile to replace Apache with Nginx. You would just need to make sure to include the relevant Nginx configuration. I might be able to take a stab at it this week if I can find some time but no promises.

If i was going to do it, i'd do this. but outside of apache and iis i'm not familiar with the other webservers

**Comment by JOJ0** [comment]
Thanks so much, that is all very great advice! Keep me posted if you find the time to test something. I hope I manage to come back to this project soon. Thanks again!

**Comment by JOJ0** [comment]
> According to this [SO post](https://stackoverflow.com/questions/66617147/how-to-fix-38function-not-implemented-ah00141-could-not-initialize-random-n), the last working version of Apache that didn't require this kernel feature is 2.4.18.
> 
> You could modify the Ampache Dockerfile to manually install that version of Apache. This isn't a great solution as it requires you to run vulnerable software.

I couldn't let it be and thought, since my Ampache won't be accessible on the internet anyway, this could be the quickest solution, but I quickly ran out knowledge about current possibilities with Debian package management.

Wasn't it always the case that getting an older version of a Debian package involves enabling the repo in /etc/apt/sources.list.d/ for older stable version, in that case Debian Buster, to get access to that specific package version?

Is there maybe new possibilities like "snapshot.debian.org", I'm not sure if I recall this service from my days back when I had more Debian admin knowledge. 

Maybe before I dig to deep, the straight up question, is there a quick way to install 2.4.18 Apache? What to roughly change in the Dockerfile? (building the container takes ages on my machine, thus it's pretty frustrating if I trial and error what to change exactly!)

Thanks!!!!!!!!!
